crates/circuit/src/circuit_data.rs

```diff
@@ -1135,6 +1135,29 @@ impl CircuitData {
         Ok(())
     }
 
+    /// Append a packed operation to this CircuitData
+    pub fn push_packed_operation(
+        &mut self,
+        operation: PackedOperation,
+        params: &[Param],
+        qargs: &[Qubit],
+        cargs: &[Clbit],
+    ) -> PyResult<()> {
+        let params = (!params.is_empty()).then(|| Box::new(params.iter().cloned().collect()));
+        let qubits = self.qargs_interner.insert(qargs);
+        let clbits = self.cargs_interner.insert(cargs);
+        self.data.push(PackedInstruction {
+            op: operation,
+            qubits,
+            clbits,
+            params,
+            extra_attrs: ExtraInstructionAttributes::default(),
+            #[cfg(feature = "cache_pygates")]
+            py_op: OnceLock::new(),
+        });
+        Ok(())
+    }
+
```

Comment:
I needed this function in addition to the already existing function `push_standard_gate`, but the code is literally identical. I am wondering if these two versions can be somehow combined?

Reply:
I think the difference is that `push_standard_gate is optimized a bit for gates. We know that a gate has no clbits so we don't bother calling `cargs_interner.insert()`. The idea was to save the overhead of having to repeatedly query the interner when we know ahead of time that it's going to be the same default value for an empty list. 

There is probably a path we could add to unify the functions, maybe change the method internals to match based on the type of  `operation`  so we avoid the hash lookup if it's a `StandardGate`, `PyGate`, or `UnitaryGate`. But maybe look into that in a separate PR to avoid too much complexity here since this is already pretty large.